x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RealmsAuthenticator.java

```diff
@@ -60,17 +59,15 @@ public String name() {
     public AuthenticationToken extractCredentials(Context context) {
         final AuthenticationToken authenticationToken = extractToken(context);
         if (authenticationToken != null) {
-            authenticationTokenExtracted = true;
+            // Once a token is extracted by realms, from the thread context,
+            // authentication must not handle the null-token case (in case no realm can verify the extracted token).
+            // In other words, the handle null-token case (i.e. authenticate as the anonymous user) runs only when no realm can extract
+            // a token from the thread context (i.e. from the request).
+            context.setHandleNullToken(false);
         }
         return authenticationToken;
     }
 
-    @Override
-    public boolean canBeFollowedByNullTokenHandler() {
-        // TODO: once a token is extracted by realms, we should no longer handle null token if no realm can authenticate the token
-        return false == authenticationTokenExtracted;
-    }
```

Comment:
This is one of the important changes in this PR, that simplifies the `AuthenticatorChain` logic.

Reply:
@jakelandis I actually think we can remove the `context.setHandleNullToken(false);` statement altogether here.

If the `RealmsAuthenticator#extractCredentials` returns non-`null` here, `RealmsAuthenticator#authenticate` is going to be invoked with that extracted token, and that method would never return `AuthenticationResult.Status.CONTINUE` (e.g. https://github.com/elastic/elasticsearch/blob/fc5c2d8695032c63aee44c804b241b643bc35a2f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RealmsAuthenticator.java#L273), which is the precondition for handling the null token (i.e. https://github.com/elastic/elasticsearch/blob/fc5c2d8695032c63aee44c804b241b643bc35a2f/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticatorChain.java#L124).

I only recently realized this.
I prefer I handle it in a self-contained follow-up.